text/2023-04-26-color-outlines.md

```diff
@@ -0,0 +1,198 @@
+
+## Interface
+
+This is a rough outline of the proposed interface which would introduce a new
+`ColorPen` trait for accessing color outlines.
+
+### Supporting types
+
+Basic types that are used by the new `ColorPen` trait.
+
+```rust
+/// Extents of a rectangular region.
+pub struct BoundingBox;
+
+/// Affine transformation matrix.
+pub struct Transform;
+
+/// 32-bit RGBA unpremultiplied color.
+pub struct Color;
+
+/// Enumeration containing data for solid or gradient fills.
+pub enum Brush<'a>;
+
+/// Contains a glyph path for clips or fills. Offers a method that can feed
+/// path commands to a Pen.
+pub struct Path<'a>;
+```
+
+### Scaler methods
+
+The `Scaler` type will gain two new methods:
+
+```rust
+fn has_color_outlines(&self) -> bool
+```
+Simply returns true if a `COLR` table is present.
+
+```rust
+fn color_outline(
+    &self, 
+    glyph_id: GlyphId, 
+    palette_fn: impl Fn(u16) -> Option<Color>, 
+    pen: &mut impl ColorPen
+) -> Result<(), Error>
+```
+Loads a color outline for the specified glyph and invokes the appropriate
+callbacks on the pen.
+
+The `palette_fn` parameter is used to specify the mapping from palette
+color index to a color value. This can be used to implement overrides and
+a helper will also be provided for constructing this function for basic
+palette support.
+
+### The `ColorPen` trait
+
+This trait contains methods that are implemented by the caller to access
+the sequence of operations necessary to render a color outline.
+
+> Note: When provided below, the code in block quotes demonstrates how these
+operations might roughly be implemented for a Skia based renderer (assuming
+a local variable `canvas` of type `SkCanvas`)
+
+#### Methods
+
+```rust
+fn bounds(&mut self, bounds: BoundingBox)
+```
+This is guaranteed to be the first method called on the pen to allow an
+appropriately sized surface to be allocated if necessary. It will pass the
+value of the `ClipBox` if available, otherwise a bounding box computed
+during evaluation of the paint graph.
+
+In accordance with the spec, if the paint graph is determined to be
+unbounded, no methods on the pen will be called and the `color_outline`
+method will return an error.
+
+```rust
+fn push_transform(&mut self, transform: Transform)
+```
+Push a transform onto the state stack.
+
+> `canvas->save();`<br>
+`canvas->concat(TransformToSkMatrix(transform));`
+
+```rust
+fn pop_transform(&mut self)
+```
+Pop a transform from the state stack.
+
+> `canvas->restore();`<br>
+
+```rust
+fn push_clip(&mut self, glyph_id: Glyph_id, path: &Path)
+```
+Push a clip path onto the state stack.
+
+> `SkPath clipPath = PathToSkPath(path);`<br>
+`canvas->save();`<br>
+`canvas->clip(clipPath, true);`
+
+```rust
+fn pop_clip(&mut self)
+```
+Pop a clip path from the state stack.
+
+> `canvas->restore();`<br>
+
+```rust
+fn push_layer(&mut self, mode: Option<CompositeMode>)
+```
+Push a new layer to the state stack, optionally with a composite mode.
```

Comment:
Might make sense to describe here which side of the COLRv1 graph this traverses / which layer becomes which: I.e. what's `sourcePaint` what's backdrop paint and this usually maps a `PaintComposite` operation.

Reply:
Great point! I’ll add some notes about exactly how `PaintComposite` maps to a sequence of callbacks.